### PR TITLE
Add examples on plugins usage

### PR DESCRIPTION
**Pull request by olga24912**
In the `examples` folder you can find the basic usage example for each plugin. The examples are presented both as README.md and as tests for the contracts.

### CODE REVIEW

File: near-plugins/src/upgradable.rs [hunk 1]
```diff
@@ -24,7 +24,7 @@ use near_sdk::{AccountId, CryptoHash, Promise};
 use serde::Serialize;
 
 pub trait Upgradable {
-    /// Key of storage slot to save the current owner.
+    /// Key of storage slot to save the stage code.
```
Comment (sept-en) [inline near-plugins/src/upgradable.rs:27]:
```suggestion
    /// Key of storage slot to save the staged code.
```

File: examples/upgradable-examples/upgradable_base/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    pub fn inc2(&mut self) {
+        self.counter += 2;
+    }
+
+    pub fn get_counter(&self) -> u64 {
+        self.counter
+    }
+}
```
Comment (sept-en) [inline examples/upgradable-examples/upgradable_base/src/lib.rs:56]: There's no need to create 2 separate structs and implementations that are the same. The more clean approach would be to have one `Counter`  contract and hide `inc2()` and/or `inc1()` by the feature flag. Or maybe even having just one `inc()` method, which changes the incrementation value depending on the feature.

File: examples/upgradable-examples/upgradable_base/build.sh [hunk 3]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
+
+RUSTFLAGS='-C link-arg=-s' cargo build --features "counter2" --target wasm32-unknown-unknown --release
+mv  ../../target/wasm32-unknown-unknown/release/upgradable_base.wasm ../../target/wasm32-unknown-unknown/release/upgradable_base_second.wasm
+RUSTFLAGS='-C link-arg=-s' cargo build --features "counter1" --target wasm32-unknown-unknown --release
```
Comment (sept-en) [inline examples/upgradable-examples/upgradable_base/build.sh:5]: It's better to store both files in the `examples/res` directory.

File: examples/upgradable-examples/README.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+1. Generate `wasm` file by running `build.sh` script. The target file will be `../target/wasm32-unknown-unknown/release/upgradable_base.wasm`
+2. Run tests `cargo test`
+
+```shell
+$ cd upgradable_base
+$ ./build.sh
+$ cargo test
+```
+
+For tests, we use `workspaces` library and `sandbox` environment for details you can explorer `../near-plugins-test-utils` crate.
```
Comment (sept-en) [inline examples/upgradable-examples/README.md:223]:
```suggestion
For tests, we use `workspaces` library and `sandbox` environment. For details, you can explore `../near-plugins-test-utils` crate.
```

File: examples/upgradable-examples/upgradable_base/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        assert!(call_borsh_arg(&contract, "up_stage_code", wasm).await);
+        assert!(call!(contract, "up_deploy_code").await);
+        check_counter(&contract, 1).await;
+
+        assert!(!call!(contract, "inc1").await);
+        check_counter(&contract, 1).await;
+
+        assert!(call!(contract, "inc2").await);
+        check_counter(&contract, 3).await;
+    }
```
Comment (sept-en) [inline examples/upgradable-examples/upgradable_base/src/lib.rs:109]:
```suggestion
    async fn call_method_with_borsh_args(contract: &Contract, method_name: &str, args: Vec<u8>) -> bool {
        contract.call(method_name)
                .args_borsh(args)
                .max_gas()
                .transact()
        .await.unwrap().is_success()
    }

    //https://docs.near.org/sdk/rust/promises/deploy-contract
    #[tokio::test]
    async fn base_scenario() {
        let (_, contract) = get_contract().await;
        assert!(call!(contract,"new").await);

        assert!(call!(contract, "inc1").await);
        check_counter(&contract, 1).await;

        assert!(!call!(contract, "inc2").await);
        check_counter(&contract, 1).await;

        let wasm = std::fs::read(WASM_FILEPATH_SECOND).unwrap();

        assert!(call_method_with_borsh_args(&contract, "up_stage_code", wasm).await);
        assert!(call!(contract, "up_deploy_code").await);
        check_counter(&contract, 1).await;

        assert!(!call!(contract, "inc1").await);
        check_counter(&contract, 1).await;

        assert!(call!(contract, "inc2").await);
        check_counter(&contract, 3).await;
    }
```
```suggestion
    async fn call_borsh_arg(contract: &Contract, method_name: &str, args: Vec<u8>) -> bool {
        contract.call(method_name)
                .args_borsh(args)
                .max_gas()
                .transact()
        .await.unwrap().is_success()
    }

    //https://docs.near.org/sdk/rust/promises/deploy-contract
    #[tokio::test]
    async fn base_scenario() {
        let (_, contract) = get_contract().await;
        assert!(call!(contract,"new").await);

        assert!(call!(contract, "inc1").await);
        check_counter(&contract, 1).await;

        assert!(!call!(contract, "inc2").await);
        check_counter(&contract, 1).await;

        let wasm = std::fs::read(WASM_FILEPATH_SECOND).unwrap();

        assert!(call_borsh_arg(&contract, "up_stage_code", wasm).await);
        assert!(call!(contract, "up_deploy_code").await);
        check_counter(&contract, 1).await;

        assert!(!call!(contract, "inc1").await);
        check_counter(&contract, 1).await;

        assert!(call!(contract, "inc2").await);
        check_counter(&contract, 3).await;
    }
```

**Review by sept-en — CHANGES_REQUESTED** [review]

File: examples/upgradable-examples/upgradable_base/src/lib.rs (hunk 2 above)
Comment (olga24912) [reply examples/upgradable-examples/upgradable_base/src/lib.rs]: Fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/f5b3b05e9f8c9e00acc45a5ed6d119d1d8d3c553

File: examples/upgradable-examples/upgradable_base/build.sh (hunk 3 above)
Comment (olga24912) [reply examples/upgradable-examples/upgradable_base/build.sh:5]: If you run the examples/build.sh script the both files will be stored in examples/res directory. 

examples/build.sh runs all build.sh scripts and copy all the wasm files into res directory. 

Do you think I should copy wasm files to res dir in all build.sh scripts?

File: examples/access-controllable-examples/README.md [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+An access control mechanism that allows you to specify which groups of users can access certain functions.
+
+```rust
+use near_plugins::AccessControllable;
+use near_plugins::AccessControlRole;
+use near_plugins_derive::access_control;
+use near_plugins_derive::access_control_any;
+use near_sdk::near_bindgen;
+use borsh::{BorshSerialize, BorshDeserialize};
+use near_plugins::events::AsEvent;
```
Comment (mooori) [inline examples/access-controllable-examples/README.md:12]:
```suggestion
```

This shouldn't be required anymore (ref. #21).

File: examples/access-controllable-examples/README.md [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+```shell
+$ near view <CONTRACT_ACCOUNT> acl_is_super_admin '{"account_id": "<SUPER_ADMIN_ACCOUNT>" }' 
+View call: <CONTRACT_ACCOUNT>.acl_is_super_admin({"account_id": "<SUPER_ADMIN_ACCOUNT>"})
+true
+```
+
+### acl_add_admin
+`acl_add_admin` is a method that adds a new admin for a specific group.
+Admins' rights don't allow running group-specific functions, but group admins can control the group member list.
+This method can be run by the admin of a specific group or by the super admin.
```
Comment (mooori) [inline examples/access-controllable-examples/README.md:101]:
```suggestion
This method can be run by an admin of a specific group or by a super admin.
```

There might be multiple admins and super admins.

**Review by mooori — COMMENTED** [review]
It might be mentioned that plugins emit events. They can be found by searching for types that implement the trait [`AsEvent`](https://github.com/aurora-is-near/near-plugins/blob/a958a307a75a944c0ad2db569662abd7a39044c0/near-plugins/src/events.rs#L21-L34). Then there should be at least one place where the `emit` method of these types is called, for example:

https://github.com/aurora-is-near/near-plugins/blob/23977fa943b77466a6eef8d8e6bcf4ba53252c8f/near-plugins-derive/src/access_controllable.rs#L147-L151

---

I left some comments in the files related to Acl. In addition:

- There might be multiple admins per role/group and there can even be more than one super admin. The last part is subtle since on the trait there's only `AccessControllable::acl_init_super_admin`.

```rust
// Assume we're inside the contract's #[init] function.
// Besides self.acl_init_super_admin there's also:
self.__acl.add_super_admin_unchecked(&alice_account_id);
self.__acl.add_super_admin_unchecked(&bob_account_id); // can be called multiple times
```

- It could be mentioned that the trait [`AccessControllable`](https://github.com/aurora-is-near/near-plugins/blob/5579de43c28984d00fa9aebc0a8e755258bba6c2/near-plugins/src/access_controllable.rs#L18) offers documentation of return values. For instance `acl_add_admin` returns `Some(true)` if the operation was successful (caller had sufficient permissions).

File: examples/upgradable-examples/upgradable_base/build.sh (hunk 3 above)
Comment (sept-en) [reply examples/upgradable-examples/upgradable_base/build.sh:5]: Yes, if you use this script, it should copy to `res` dir. The top-level script could rather execute each example script. This way, you can have a more clean approach for the `Upgradable` contract, for example, which will utilize the file from the `res` directory.

Comment (olga24912) [reply examples/upgradable-examples/upgradable_base/build.sh:7]: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/5c8f7548888eaa6f0c2d2cb10b28604a1e9c9c49

File: examples/access-controllable-examples/README.md [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+
+/// All types of access groups
+#[derive(AccessControlRole, Clone, Copy)]
+pub enum UsersGroups {
+    GroupA,
+    GroupB,
+}
+
+#[near_bindgen]
+#[access_control(role_type="UsersGroups")]
```
[the remark this answers is not in the document]
Comment (olga24912) [reply examples/access-controllable-examples/README.md]: fix https://github.com/aurora-is-near/near-plugins/pull/34/commits/a334a44069b91318440e1a87c74f219a248b9920

File: examples/access-controllable-examples/README.md [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+    /// In the constructor we set up a super admin, 
+    /// which can control the member lists of all user groups
+    #[init]
+    pub fn new() -> Self {
+        let mut contract: Counter = Self{
+            counter: 0,
+            __acl: __Acl::default(),
+        };
+
+        contract.__acl.init_super_admin(&near_sdk::env::predecessor_account_id());
```
[the remark this answers is not in the document]
Comment (olga24912) [reply examples/access-controllable-examples/README.md]: fix https://github.com/aurora-is-near/near-plugins/pull/34/commits/a334a44069b91318440e1a87c74f219a248b9920

File: examples/access-controllable-examples/README.md (hunk 7 above)
Comment (olga24912) [reply examples/access-controllable-examples/README.md]: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/6fd1a4a364b2180df630ee7c4a85b3718c14c89d

File: examples/access-controllable-examples/README.md [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+
+```shell
+$ near call <CONTRACT_ACCOUNT> acl_renounce_admin '{"role": "GroupA"}' --accountId <ALICE_ACCOUNT>
+```
+
+After calling that method, Alice will not have the admin right for GroupA anymore.
+
+### acl_revoke_role
+`acl_revoke_role` is a method that removes the specified account from the list of the group member.
+Only the group admin or super admin can execute this function.
```
[the remark this answers is not in the document]
Comment (olga24912) [reply examples/access-controllable-examples/README.md]: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/6fd1a4a364b2180df630ee7c4a85b3718c14c89d

File: examples/access-controllable-examples/README.md [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+`acl_add_admin` is a method that adds a new admin for a specific group.
+Admins' rights don't allow running group-specific functions, but group admins can control the group member list.
+This method can be run by the admin of a specific group or by the super admin.
+
+```shell
+$ near call <CONTRACT_ACCOUNT> acl_add_admin '{"role": "GroupA", "account_id": "<ALICE_ACCOUNT>"}' --accountId <GROUP_A_ADMIN_ACCOUNT>
+```
+
+### acl_is_admin
+`acl_is_admin` is a _view_ method that checks if the account has an admin right for the specified group. For super admin, it will return true.
```
[the remark this answers is not in the document]
Comment (olga24912) [reply examples/access-controllable-examples/README.md]: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/6fd1a4a364b2180df630ee7c4a85b3718c14c89d

File: examples/access-controllable-examples/access_controllable_base/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+    /// In the constructor we set up a super admin,
+    /// which can control the member lists of all user groups
+    #[init]
+    pub fn new() -> Self {
+        let mut contract: Counter = Self{
+            counter: 0,
+            __acl: __Acl::default(),
+        };
+
+        contract.__acl.init_super_admin(&near_sdk::env::predecessor_account_id());
```
[the remark this answers is not in the document]
Comment (olga24912) [reply examples/access-controllable-examples/access_controllable_base/src/lib.rs]: fixed https://github.com/aurora-is-near/near-plugins/pull/34/commits/a334a44069b91318440e1a87c74f219a248b9920

File: examples/access-controllable-examples/README.md (hunk 6 above)
Comment (olga24912) [reply examples/access-controllable-examples/README.md]: fixed https://github.com/aurora-is-near/near-plugins/pull/34/commits/a334a44069b91318440e1a87c74f219a248b9920

File: examples/access-controllable-examples/README.md [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+    pub fn get_counter(&self) -> u64 {
+        self.counter
+    }
+}
+```
+
+## The contract methods description
+### acl_storage_prefix
+`acl_storage_prefix` is a method that returns the common prefix of keys for storing the members and the admins of groups.
+`__acl` by default.
```
[the remark this answers is not in the document]
Comment (olga24912) [reply examples/access-controllable-examples/README.md:72]: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/c11247d4f3fc1f49a14fd6560235b8b3b22b4751

File: examples/access-controllable-examples/README.md [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+
+### acl_has_any_role
+`acl_has_any_role` is a _view_ method for checking if an account is a member of at least one of the specified groups.
+
+```shell
+$ near view <CONTRACT_ACCOUNT> acl_has_any_role '{"roles": ["GroupA", "GroupB"], "account_id": "<ALICE_ACCOUNT>"}'
+View call: <CONTRACT_ACCOUNT>.acl_has_any_role({"roles": ["GroupA", "GroupB"], "account_id": "<ALICE_ACCOUNT>"})
+true
+```
+
```
[the remark this answers is not in the document]
Comment (olga24912) [reply examples/access-controllable-examples/README.md:192]: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/6ce74158b38e3243b03abbe8b25d827601c1547b

File: examples/access-controllable-examples/README.md [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+Doing account.functionCall()
+Transaction Id <TRANSACTION_ID>
+To see the transaction in the transaction explorer, please open this url in your browser
+https://explorer.testnet.near.org/transactions/<TRANSACTION_ID>
+[ 95, 95, 97, 99, 108 ]
+$ python3
+>>> print(' '.join(str(b) for b in bytes("__acl", 'utf8')))
+95 95 97 99 108
+```
+
```
[the remark this answers is not in the document]
Comment (olga24912) [reply examples/access-controllable-examples/README.md:86]: fixed: https://github.com/aurora-is-near/near-plugins/pull/34/commits/6ce74158b38e3243b03abbe8b25d827601c1547b

[the remark this answers is not in the document]
Comment (olga24912) [reply examples/access-controllable-examples/README.md:86]: @mooori is it possible to remove the super admin somehow?

[the remark this answers is not in the document]
Comment (mooori) [reply examples/access-controllable-examples/README.md:86]: There is [`revoke_super_admin_unchecked`](https://github.com/aurora-is-near/near-plugins/blob/5579de43c28984d00fa9aebc0a8e755258bba6c2/near-plugins-derive/src/access_controllable.rs#L169-L171), it can be called like this from inside a contract method:

```rust
self.__acl.revoke_super_admin_unchecked(&account_id);
```

**Review by karim-en — CHANGES_REQUESTED** [review]

File: examples/upgradable-examples/up_stage_code/src/main.rs [hunk 16]
```diff
@@ -0,0 +1,23 @@
+use serde_json::json;
+use tokio::runtime::Runtime;
+use workspaces::{Account, Contract};
```
[the remark this answers is not in the document]
Comment (olga24912) [reply examples/upgradable-examples/up_stage_code/src/main.rs]: fixed https://github.com/aurora-is-near/near-plugins/pull/34/commits/662b8af60d75000679333cc9e0694ecce2d50efe

File: examples/upgradable-examples/up_stage_code/src/main.rs [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+use std::env;
+
+const WASM_FILEPATH_SECOND: &str = "../../target/wasm32-unknown-unknown/release/upgradable_base_second.wasm";
+
+fn main() {
```
[the remark this answers is not in the document]
Comment (olga24912) [reply examples/upgradable-examples/up_stage_code/src/main.rs]: fixed https://github.com/aurora-is-near/near-plugins/pull/34/commits/b8b88ae3b5e8b5b65f2ac101f79119d30b68122f

File: examples/upgradable-examples/up_stage_code/src/main.rs [hunk 18]
```diff
@@ -0,0 +1,23 @@
...
+    let args: Vec<String> = env::args().collect();
+    let rt = Runtime::new().unwrap();
+    let worker = rt.block_on(workspaces::testnet()).unwrap();
```
[the remark this answers is not in the document]
Comment (olga24912) [reply examples/upgradable-examples/up_stage_code/src/main.rs]: fixed https://github.com/aurora-is-near/near-plugins/pull/34/commits/c69044ea4429608971283fed604d7ffa65d9e91c

File: examples/upgradable-examples/up_stage_code/src/main.rs [hunk 19]
```diff
@@ -0,0 +1,23 @@
+use serde_json::json;
+use tokio::runtime::Runtime;
+use workspaces::{Account, Contract};
+use std::env;
+
+const WASM_FILEPATH_SECOND: &str = "../../target/wasm32-unknown-unknown/release/upgradable_base_second.wasm";
```
[the remark this answers is not in the document]
Comment (olga24912) [reply examples/upgradable-examples/up_stage_code/src/main.rs]: fixed https://github.com/aurora-is-near/near-plugins/pull/34/commits/c69044ea4429608971283fed604d7ffa65d9e91c

**Review by mfornet — APPROVED** [review]
Thanks for writing these examples @olga24912.

**Review by karim-en — APPROVED** [review]

File: examples/access-controllable-examples/README.md [hunk 20]
```diff
@@ -0,0 +1,450 @@
...
+$ near call <CONTRACT_ACCOUNT> level_a_incr --accountId <CONTRACT_ACCOUNT>
+ERROR
+$ near view get_counter
+2
+```
+
+### A few super admins
+There may be two or more super admins. 
+By default, there are no interfaces that allow you to add or remove super admins. 
+You can add the corresponding functions to the contract as follows:
```
Comment (mooori) [inline examples/access-controllable-examples/README.md:417]: I'd rephrase it a bit, something along those lines:

```suggestion
### Multiple super admins
There may be multiple super admins. 
By default, only `acl_init_super_admin` is exposed on the contract. To add more super admins or to remove super admins, internal functions can be used. 
To make this functionality publicly available, you could add functions like the following to your contract:
```

File: examples/access-controllable-examples/README.md [hunk 21]
```diff
@@ -0,0 +1,450 @@
...
+
+```rust
+/// method for adding new super admin
+pub fn add_super_admin(&mut self, new_super_admin_account_id: &AccountId) {
+    ::near_sdk::require!(self.__acl.is_super_admin(&near_sdk::env::predecessor_account_id()),
```
Comment (mooori) [inline examples/access-controllable-examples/README.md:422]: The trait method is documented, so I'd suggest calling that:

```suggestion
    ::near_sdk::require!(self.acl_is_super_admin(near_sdk::env::predecessor_account_id()),
```

File: examples/access-controllable-examples/README.md [hunk 22]
```diff
@@ -0,0 +1,450 @@
...
+                         "Method can be run only by super admin");
+    self.__acl.add_super_admin_unchecked(new_super_admin_account_id); 
+}
+
+/// method for removing super admin
+pub fn remove_super_admin(&mut self, super_admin_account_id: &AccountId) {
+   ::near_sdk::require!(self.__acl.is_super_admin(&near_sdk::env::predecessor_account_id()),
```
Comment (mooori) [inline examples/access-controllable-examples/README.md:429]:
```suggestion
   ::near_sdk::require!(self.acl_is_super_admin(near_sdk::env::predecessor_account_id()),
```

File: examples/access-controllable-examples/README.md [hunk 23]
```diff
@@ -0,0 +1,450 @@
...
+```
+
+If the method succeeds, the following event will be emitted:
+```json
+{
+   "standard":"AccessControllable",
+   "version":"1.0.0",
+   "event":"super_admin_added",
+   "data":{
+      "account":"test.near",
```
Comment (mooori) [inline examples/access-controllable-examples/README.md:118]:
```suggestion
      "account":"<SUPER_ADMIN_ACCOUNT",
```

**Review by mooori — COMMENTED** [review]
Just some nits

File: examples/access-controllable-examples/README.md (hunk 23 above)
Comment (olga24912) [reply examples/access-controllable-examples/README.md]: fix https://github.com/aurora-is-near/near-plugins/pull/34/commits/1a7a34b88d2b4513d1de096b408ef5fc17e46209

File: examples/access-controllable-examples/README.md (hunk 22 above)
Comment (olga24912) [reply examples/access-controllable-examples/README.md]: fix https://github.com/aurora-is-near/near-plugins/pull/34/commits/1a7a34b88d2b4513d1de096b408ef5fc17e46209

File: examples/access-controllable-examples/README.md (hunk 21 above)
Comment (olga24912) [reply examples/access-controllable-examples/README.md]: fix https://github.com/aurora-is-near/near-plugins/pull/34/commits/1a7a34b88d2b4513d1de096b408ef5fc17e46209

File: examples/access-controllable-examples/README.md (hunk 20 above)
Comment (olga24912) [reply examples/access-controllable-examples/README.md]: fix https://github.com/aurora-is-near/near-plugins/pull/34/commits/1a7a34b88d2b4513d1de096b408ef5fc17e46209

**Review by mooori — APPROVED** [review]

**Review by sept-en — APPROVED** [review]

**Review by mrLSD — APPROVED** [review]
